File: src/lib_nim/music.nim

import math

type 
  NoteName* = enum
    C, Cx, Db, D, Dx, Eb, E, F, Fx, Gb, G, Gx, Ab, A, Ax, Bb, B

  Note* = tuple[name: NoteName, octave: int]

func toMidi* (note: Note): int = 
  const sharps = [C, Cx, D, Dx, E, F, Fx, G, Gx, A, Ax, B]
  const flats = [C, Db, D, Eb, E, F, Gb, G, Ab, A, Bb, B]
  var toneIndex = sharps.find(note.name)
  if toneIndex == -1:
    toneIndex = flats.find(note.name)
  (note.octave + 1) * 12 + toneIndex

func toFrequency* (midi: int): float = 
  pow(2, float(midi - 69) / 12) * 440.0

func toFrequency* (note: Note): float = 
  note.toMidi.toFrequency

func harmonicMajorScale* (baseFrequency: float): seq[float] =
  @[
    baseFrequency / 1.0,
    (baseFrequency * 7) / 6,
    (baseFrequency * 5) / 4,
    (baseFrequency * 4) / 3,
    (baseFrequency * 3) / 2,

    ((baseFrequency * 3) / 2 * 7) / 6,
    ((baseFrequency * 3) / 2 * 5) / 4
  ]

# midi chords

type 
  Chord* = enum
    maj, min, dim, maj7, d7, min7, dim7, aug, dimFull

func midiOffsets* (name: Chord): seq[int] =
  case name
    of maj: @[0, 4, 7]
    of min: @[0, 3, 7]
    of dim: @[0, 3, 6]
    of aug: @[0, 4, 8]
    of maj7: @[0, 4, 7, 11]
    of d7: @[0, 4, 7, 10]
    of min7: @[0, 3, 7, 10]
    of dim7: @[0, 3, 6, 10]
    of dimFull: @[0, 3, 6, 9]

func chord* (baseNoteMidi: int, chordName: Chord): seq[int] =
  for n in chordName.midiOffsets:
    result.add(baseNoteMidi + n)

func chord* (baseNote: Note, chordName: Chord): seq[int] =
  chord(baseNote.toMidi, chordName)

func invertUp* (chord: seq[int]): seq[int] =
  chord[1..^1] & @[chord[0] + 12]

func invertDown* (chord: seq[int]): seq[int] =
  @[chord[-1] - 12] & chord[0..^1]

func invert* (chord: seq[int], times = 1): seq[int] =
  result = chord
  if times < 0:
    for i in 1..abs(times):
      result = result.invertDown
  else:
    for i in 1..times:
      result = result.invertUp

func reverse* (chord: seq[int]): seq[int] =
  var i = chord.high
  while i >= chord.low:
    result.add(chord[i])
    i.dec

# midi scales

type
  Scale* = enum
    major, minor, harmonicMinor, majorPentatonic, minorPentatonic, blues

func midiOffsets* (scale: Scale): seq[int] =
  case scale
    of major: @[0, 2, 4, 5, 7, 9, 11]
    of minor: @[0, 2, 3, 5, 7, 8, 10]
    of harmonicMinor: @[0, 2, 3, 5, 7, 8, 11]
    of majorPentatonic: @[0, 2, 4, 7, 9]
    of minorPentatonic: @[0, 3, 5, 7, 10]
    of blues: @[0, 3, 5, 6, 7, 10]

func scale* (baseNoteMidi: int, scale: Scale): seq[int] =
  let offsets = scale.midiOffsets
  for j in 0..<3:
    for i in offsets:
      result.add(i + baseNoteMidi + (j * 12))

func scale* (baseNote: Note, scaleType: Scale): seq[int] =
  baseNote.toMidi.scale(scaleType)

func sharp* (note: int): int = 
  note + 1

func flat* (note: int): int =
  note - 1

File: src/lib/music.ts
export const midiToFc = (midi: number) => {
	return 2 ** ((midi - 69) / 12) * 440
}

export type NoteName =
	| 'C'
	| 'C#'
	| 'Db'
	| 'D'
	| 'D#'
	| 'Eb'
	| 'E'
	| 'F'
	| 'F#'
	| 'Gb'
	| 'G'
	| 'G#'
	| 'Ab'
	| 'A'
	| 'A#'
	| 'Bb'
	| 'B'

export type NoteValue = `${NoteName}-${number}`

export type Note = {
	note: NoteName
	octave: number
}

function noteFromValue(value: NoteValue): Note {
	const [note, octave] = value.split('-')
	return {
		note: note as NoteName,
		octave: parseInt(octave),
	}
}

export const noteToMidi = (note: Note | NoteValue) => {
	const n = typeof note === 'string' ? noteFromValue(note) : note
	const { note: name, octave } = n
	const sharps: NoteName[] = [
		'C',
		'C#',
		'D',
		'D#',
		'E',
		'F',
		'F#',
		'G',
		'G#',
		'A',
		'A#',
		'B',
	]
	const flats: NoteName[] = [
		'C',
		'Db',
		'D',
		'Eb',
		'E',
		'F',
		'Gb',
		'G',
		'Ab',
		'A',
		'Bb',
		'B',
	]
	let toneIndex = sharps.indexOf(name)
	if (toneIndex === -1) {
		toneIndex = flats.indexOf(name)
	}
	return (octave + 1) * 12 + toneIndex
}

export type ChordName = 'maj' | 'min' | 'maj7' | '7' | 'min7' | 'dim' | 'dim7'

export function chord(baseNote: number | NoteValue, chordName: ChordName) {
	const base = typeof baseNote === 'number' ? baseNote : noteToMidi(baseNote)
	const chord = chordName.includes('min')
		? [0, 3, 7]
		: chordName.includes('dim')
			? [0, 3, 6]
			: [0, 4, 7]
	if (chordName === 'maj7') {
		chord.push(11)
	} else if (chordName.includes('7')) {
		chord.push(10)
	}

	return chord.map((n) => base + n)
}

function invertUp(chord: number[]) {
	const [first, ...rest] = chord
	return [...rest, first + 12]
}

function invertDown(chord: number[]) {
	const [last, ...rest] = [...chord].reverse()
	return [last - 12, ...rest.reverse()]
}

export function invert(chord: number[], times = 1) {
	if (times < 0) {
		times = Math.abs(times)
		for (let i = 0; i < times; i++) {
			chord = invertDown(chord)
		}
	} else {
		for (let i = 0; i < times; i++) {
			chord = invertUp(chord)
		}
	}
	return chord
}

export function harmonicMajorScale(baseFrequency: number) {
	return [
		baseFrequency / 1,

		(baseFrequency * 7) / 6,
		// (baseTone * 6) / 5, // no minor third
		(baseFrequency * 5) / 4,
		(baseFrequency * 4) / 3,
		(baseFrequency * 3) / 2,

		(((baseFrequency * 3) / 2) * 7) / 6,
		(((baseFrequency * 3) / 2) * 5) / 4,
	]
}

export enum ScaleType {
	MAJOR = 'major',
	MINOR = 'minor',
	HARMONIC_MINOR = 'harmonicMinor',
}

export function scaleIntervals(type: ScaleType) {
	switch (type) {
		case ScaleType.MAJOR:
			return [0, 2, 4, 5, 7, 9, 11]
		case ScaleType.MINOR:
			return [0, 2, 3, 5, 7, 8, 10]
		case ScaleType.HARMONIC_MINOR:
			return [0, 2, 3, 5, 7, 8, 11]
	}
}

export function scale(baseNote: NoteValue, type: ScaleType) {
	const intervals = scaleIntervals(type)
	const notes = intervals
		.map((i) => i + noteToMidi(baseNote))
		.concat(intervals.map((i) => i + noteToMidi(baseNote) + 12))
		.concat(intervals.map((i) => i + noteToMidi(baseNote) + 24))

	return function (value: number, prefix?: '#' | 'b') {
		const note = notes[value - 1]
		if (prefix === '#') {
			return note + 1
		} else if (prefix === 'b') {
			return note - 1
		}
		return note
	}
}

export const octaveTransposedScale =
	(s: ReturnType<typeof scale>, octaves = 1) =>
	(value: number, prefix?: '#' | 'b') =>
		s(value, prefix) + 12 * octaves
